packages/pool/contracts/TransferUtils.sol

```diff
@@ -6,43 +6,141 @@ import "./interfaces/ITransferUtils.sol";
+    /// `getUserLocked()` call gas cost may exceed the block gas limit. In that
+    /// case, the user may call this method multiple times to have their locked
+    /// tokens calculated.
+    /// @param userAddress User address
+    /// @param noEpochsPerIteration Number of epochs per iteration
+    /// @return finished Calculation has finished in this call
+    function calculateUserLockedIteratively(
+        address userAddress,
+        uint256 noEpochsPerIteration
+        )
+        external
+        override
+        returns (bool finished)
+    {
+        require(noEpochsPerIteration > 0, "Iteration window invalid");
+        payReward();
+        Checkpoint[] storage _userShares = users[userAddress].shares;
+        uint256 userSharesLength = _userShares.length;
+        require(userSharesLength != 0, "User never had shares");
+        uint256 currentEpoch = block.timestamp / EPOCH_LENGTH;
+        LockedCalculationState storage state = userToLockedCalculationState[userAddress];
+        // Reset the state if there was no calculation made in this epoch
+        if (state.initialIndEpoch != currentEpoch)
+        {
+            state.initialIndEpoch = currentEpoch;
+            state.nextIndEpoch = currentEpoch;
+            state.locked = 0;
+        }
+        uint256 indEpoch = state.nextIndEpoch;
+        uint256 locked = state.locked;
+        uint256 oldestLockedEpoch = currentEpoch - REWARD_VESTING_PERIOD > genesisEpoch
+            ? currentEpoch - REWARD_VESTING_PERIOD + 1
+            : genesisEpoch + 1;
+        for (; indEpoch >= oldestLockedEpoch; indEpoch--)
+        {
+            if (state.nextIndEpoch >= indEpoch + noEpochsPerIteration)
+            {
+                state.nextIndEpoch = indEpoch;
+                state.locked = locked;
+                emit CalculatingUserLocked(
+                    userAddress,
+                    indEpoch,
+                    oldestLockedEpoch
+                    );
+                return false;
+            }
+            Reward storage lockedReward = epochIndexToReward[indEpoch];
+            if (lockedReward.atBlock != 0)
+            {
+                uint256 userSharesThen = getValueAtWithBinarySearch(_userShares, lockedReward.atBlock);
+                locked = locked + (lockedReward.amount * userSharesThen / lockedReward.totalSharesThen);
+            }
+        }
+        state.nextIndEpoch = indEpoch;
+        state.locked = locked;
+        emit CalculatedUserLocked(userAddress, locked);
+        return true;
+    }
+
+    /// @notice Called by the user to withdraw after their locked token amount
+    /// is calculated with repeated calls to `calculateUserLockedIteratively()`
+    /// @dev Only use `calculateUserLockedIteratively()` and this method if
+    /// `withdrawRegular()` hits the block gas limit
+    /// @param destination Token transfer destination
+    /// @param amount Amount to be withdrawn
+    function withdrawWithPrecalculatedLocked(
```

Comment:
maybe `withdrawPrecalculated` could be shorter?

Reply:
Agreed